[SPARK-10941] [SQL] Refactor AggregateFunction2 and AlgebraicAggregate interfaces to improve code clarity

JoshRosen opened the pull request:
This patch refactors several of the Aggregate2 interfaces in order to improve code clarity.

The biggest change is a refactoring of the `AggregateFunction2` class hierarchy. In the old code, we had a class named `AlgebraicAggregate` that inherited from `AggregateFunction2`, added a new set of methods, then banned the use of the inherited methods. I found this to be fairly confusing because.

If you look carefully at the existing code, you'll see that subclasses of `AggregateFunction2` fall into two disjoint categories: imperative aggregation functions which directly extended `AggregateFunction2` and declarative, expression-based aggregate functions which extended `AlgebraicAggregate`. In order to make this more explicit, this patch refactors things so that `AggregateFunction2` is a sealed abstract class with two subclasses, `ImperativeAggregateFunction` and `ExpressionAggregateFunction`. The superclass, `AggregateFunction2`, now only contains methods and fields that are common to both subclasses.

After making this change, I updated the various AggregationIterator classes to comply with this new naming scheme. I also performed several small renamings in the aggregate interfaces themselves in order to improve clarity and rewrote or expanded a number of comments.

JoshRosen committed:
Remove unused initialInputBufferOffset from TungstenAggregate.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43220 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43220/consoleFull) for   PR 8973 at commit [`37a0731`](https://github.com/apache/spark/commit/37a0731dc13da179cf701d6a4c5cd56ed1db7615).

JoshRosen commented:
Ah, I think that I may see the reason for this now: it appears that the bufferOffsets in `AggregateFunction2` are not relevant when you're dealing with `AlgebraicAggregate` instances.

To me, this is further evidence that the current inheritance hierarchy is wrong: `AlgebraicAggregate` inherits from `AggregateFunction2`, adds a new set of methods, then effectively bans the use of the inherited methods. This is really confusing. I think that it's an anti-pattern / bad code smell if you end up inheriting and wanting to remove methods inherited from the superclass.

yhuai commented:
The reason that `initialInputBufferOffset` is there is to use that when we want to support aggregate functions that do not support code-gen.

Regarding the hierarchy comment, what is your suggestion?

JoshRosen commented:
Regarding the hierarchy, I think that we should have an AggregateFunction2 class with two subclasses, one to handle aggregates which can be implemented in terms of expressions (basically, just rename AlgebraicAggregate) and another for dealing with interpreted aggregates only (named something like InterpretedAggregate).

yhuai commented:
Makes sense. What will the interface of `AggregateFunction2` look like?

JoshRosen committed:
WIP demonstration of aggregate function hierarchy refactoring.

JoshRosen commented:
@yhuai, just pushed a WIP commit demonstrating one possible refactoring. This passes the hive AggregationQuerySuites locally. I'll comment on a few points inline.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala:
```diff
@@ -95,12 +95,32 @@ private[sql] case class AggregateExpression2(
 
-abstract class AggregateFunction2
-  extends Expression with ImplicitCastInputTypes {
+sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes {
```
Logically, AggregateFunction2 is a union-type, the least-upper-bound of InterpretedAggregateFunction and ExpressionAggregateFunction (the temporary name which I'm using for expression-based aggregates).

JoshRosen replied:
I plan to write some proper Scaladoc for this class to explain the role of its methods.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala:
```diff
@@ -110,16 +118,16 @@ abstract class AggregationIterator(
     functions
   }
 
-  // Positions of those non-algebraic aggregate functions in allAggregateFunctions.
```
Throughout this file, "non-algebraic" has been replaced with "interpreted". This renaming lets us remove one level of negation in a bunch of places, simplifying a bunch of code.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/AggregationIterator.scala:
```diff
@@ -131,24 +139,26 @@ abstract class AggregationIterator(
-  // All non-algebraic aggregate functions with mode Partial, PartialMerge, or Final.
-  private[this] val nonCompleteNonAlgebraicAggregateFunctions: Array[AggregateFunction2] =
-    nonCompleteAggregateFunctions.collect {
-      case func: AggregateFunction2 if !func.isInstanceOf[AlgebraicAggregate] => func
```
For instance, here's an example of how changing "non-algebraic" to "interpreted" can simplify things a bit.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TungstenAggregationIterator.scala:
```diff
@@ -186,77 +182,71 @@ class TungstenAggregationIterator(
         (currentBuffer: UnsafeRow, row: InternalRow) => {
-          algebraicUpdateProjection.target(currentBuffer)
-          algebraicUpdateProjection(joinedRow(currentBuffer, row))
+          updateProjection.target(currentBuffer)
```
For _now_, TungstenAggregationIterator only supports expression-based aggregates, and thus we don't need to explicitly qualify that this is an expression-based update projection. When I refactor TungstenAggregationIterator to support interpreted aggregates as well, I'll re-name to disambiguate.

JoshRosen commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/SortBasedAggregate.scala:
```diff
@@ -27,6 +27,14 @@ import org.apache.spark.sql.execution.{UnsafeFixedWidthAggregationMap, SparkPlan
+/**
+ * Sort-based aggregation iterator.
+ *
+ * @param initialInputBufferOffset
```
I think that this offset should be documented higher up, as part of InterpretedAggregateFunction. I'll clean this up while I add docs.

SparkQA commented:
[Test build #43221 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43221/consoleFull) for   PR 8973 at commit [`43f211d`](https://github.com/apache/spark/commit/43f211d3b3cfe014f9774f47791b282e8dc53e09).

SparkQA commented:
[Test build #43221 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43221/console) for   PR 8973 at commit [`43f211d`](https://github.com/apache/spark/commit/43f211d3b3cfe014f9774f47791b282e8dc53e09).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregateFunction `
  * `case class Count(child: Expression) extends ExpressionAggregateFunction `
  * `case class First(child: Expression) extends ExpressionAggregateFunction `
  * `case class Last(child: Expression) extends ExpressionAggregateFunction `
  * `case class Max(child: Expression) extends ExpressionAggregateFunction `
  * `case class Min(child: Expression) extends ExpressionAggregateFunction `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregateFunction `
  * `case class Sum(child: Expression) extends ExpressionAggregateFunction `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class InterpretedAggregateFunction extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43221/
Test FAILed.

JoshRosen committed:
Fix Scalastyle

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43222 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43222/consoleFull) for   PR 8973 at commit [`8967793`](https://github.com/apache/spark/commit/89677933688d30ac8930c26f900d2691e34aec97).

SparkQA commented:
[Test build #43220 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43220/console) for   PR 8973 at commit [`37a0731`](https://github.com/apache/spark/commit/37a0731dc13da179cf701d6a4c5cd56ed1db7615).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43220/
Test PASSed.

SparkQA commented:
[Test build #43222 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43222/console) for   PR 8973 at commit [`8967793`](https://github.com/apache/spark/commit/89677933688d30ac8930c26f900d2691e34aec97).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregateFunction `
  * `case class Count(child: Expression) extends ExpressionAggregateFunction `
  * `case class First(child: Expression) extends ExpressionAggregateFunction `
  * `case class Last(child: Expression) extends ExpressionAggregateFunction `
  * `case class Max(child: Expression) extends ExpressionAggregateFunction `
  * `case class Min(child: Expression) extends ExpressionAggregateFunction `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregateFunction `
  * `case class Sum(child: Expression) extends ExpressionAggregateFunction `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class InterpretedAggregateFunction extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43222/
Test PASSed.

marmbrus commented:
Seems fine to me

JoshRosen commented:
Alright. Will focus on a pass of comments updates now, then will file a JIRA for this.

JoshRosen committed:
Merge remote-tracking branch 'origin/master' into tungsten-agg-comments

JoshRosen commented:
Per some offline discussion with Michael, I've decided to renamed InterpretedAggregateFunction to ImperativeAggregateFunction: it's logically the opposite of ExpressionBasedAggregateFunction and the new name seems a touch clearer.

JoshRosen committed:
Rename InterpretedAggregateFunction to ImperativeAggregateFunction.

JoshRosen committed:
Add a ton more documentation.

JoshRosen committed:
Cleanup of cloneBufferSchema.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43267 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43267/consoleFull) for   PR 8973 at commit [`ea92495`](https://github.com/apache/spark/commit/ea924953380d2eb0adf44bc722884ec7d964c385).

SparkQA commented:
[Test build #43267 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43267/console) for   PR 8973 at commit [`ea92495`](https://github.com/apache/spark/commit/ea924953380d2eb0adf44bc722884ec7d964c385).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregateFunction `
  * `case class Count(child: Expression) extends ExpressionAggregateFunction `
  * `case class First(child: Expression) extends ExpressionAggregateFunction `
  * `case class Last(child: Expression) extends ExpressionAggregateFunction `
  * `case class Max(child: Expression) extends ExpressionAggregateFunction `
  * `case class Min(child: Expression) extends ExpressionAggregateFunction `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregateFunction `
  * `case class Sum(child: Expression) extends ExpressionAggregateFunction `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class ImperativeAggregateFunction extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43267/
Test FAILed.

JoshRosen committed:
More interpreted -> imperative renaming.

JoshRosen committed:
One more rename

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

JoshRosen commented:
Alright, I've gone ahead and filed a JIRA, performed some more renaming, and added a _ton_ more Scaladoc to the new split interfaces in order to clarify how they're used. Reviewers, please pay attention to how I renamed some of the attribute and buffer fields in order to clarify field offset handling, etc.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala:
```diff
@@ -95,98 +92,186 @@ private[sql] case class AggregateExpression2(
-  def withNewMutableBufferOffset(newMutableBufferOffset: Int): Unit = {
-    mutableBufferOffset = newMutableBufferOffset
-  }
+  final lazy val inputAggBufferAttributes: Seq[Attribute] = aggBufferAttributes.map(_.newInstance())
```
This was duplicated in all subclasses, but I don't see any reason to let subclasses override it (seems like that will lead to bugs).

SparkQA commented:
[Test build #43268 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43268/consoleFull) for   PR 8973 at commit [`9800f64`](https://github.com/apache/spark/commit/9800f64307cfa2232a17b96198856dafcbab0dd9).

SparkQA commented:
[Test build #43268 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43268/console) for   PR 8973 at commit [`9800f64`](https://github.com/apache/spark/commit/9800f64307cfa2232a17b96198856dafcbab0dd9).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregateFunction `
  * `case class Count(child: Expression) extends ExpressionAggregateFunction `
  * `case class First(child: Expression) extends ExpressionAggregateFunction `
  * `case class Last(child: Expression) extends ExpressionAggregateFunction `
  * `case class Max(child: Expression) extends ExpressionAggregateFunction `
  * `case class Min(child: Expression) extends ExpressionAggregateFunction `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregateFunction `
  * `case class Sum(child: Expression) extends ExpressionAggregateFunction `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class ImperativeAggregateFunction extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43268/
Test FAILed.

JoshRosen committed:
Fix Scalastyle.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43271 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43271/consoleFull) for   PR 8973 at commit [`e34e22e`](https://github.com/apache/spark/commit/e34e22ef25dabcf5ee03fd51631a3d8f1a227070).

SparkQA commented:
[Test build #43271 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43271/console) for   PR 8973 at commit [`e34e22e`](https://github.com/apache/spark/commit/e34e22ef25dabcf5ee03fd51631a3d8f1a227070).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregateFunction `
  * `case class Count(child: Expression) extends ExpressionAggregateFunction `
  * `case class First(child: Expression) extends ExpressionAggregateFunction `
  * `case class Last(child: Expression) extends ExpressionAggregateFunction `
  * `case class Max(child: Expression) extends ExpressionAggregateFunction `
  * `case class Min(child: Expression) extends ExpressionAggregateFunction `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregateFunction `
  * `case class Sum(child: Expression) extends ExpressionAggregateFunction `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class ImperativeAggregateFunction extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43271/
Test FAILed.

JoshRosen committed:
Fix compilation error.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43274 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43274/consoleFull) for   PR 8973 at commit [`fd6f853`](https://github.com/apache/spark/commit/fd6f853b9e8077ed76a81392f37dbec7fe448262).

SparkQA commented:
[Test build #43274 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43274/console) for   PR 8973 at commit [`fd6f853`](https://github.com/apache/spark/commit/fd6f853b9e8077ed76a81392f37dbec7fe448262).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregateFunction `
  * `case class Count(child: Expression) extends ExpressionAggregateFunction `
  * `case class First(child: Expression) extends ExpressionAggregateFunction `
  * `case class Last(child: Expression) extends ExpressionAggregateFunction `
  * `case class Max(child: Expression) extends ExpressionAggregateFunction `
  * `case class Min(child: Expression) extends ExpressionAggregateFunction `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregateFunction `
  * `case class Sum(child: Expression) extends ExpressionAggregateFunction `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class ImperativeAggregateFunction extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43274/
Test FAILed.

JoshRosen committed:
Fix more compile breaks

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43275/consoleFull) for   PR 8973 at commit [`128f29f`](https://github.com/apache/spark/commit/128f29fa2faf7a6eb36faf4db78beb2fdbf9b17a).

SparkQA commented:
[Test build #43275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43275/console) for   PR 8973 at commit [`128f29f`](https://github.com/apache/spark/commit/128f29fa2faf7a6eb36faf4db78beb2fdbf9b17a).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregateFunction `
  * `case class Count(child: Expression) extends ExpressionAggregateFunction `
  * `case class First(child: Expression) extends ExpressionAggregateFunction `
  * `case class Last(child: Expression) extends ExpressionAggregateFunction `
  * `case class Max(child: Expression) extends ExpressionAggregateFunction `
  * `case class Min(child: Expression) extends ExpressionAggregateFunction `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregateFunction `
  * `case class Sum(child: Expression) extends ExpressionAggregateFunction `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class ImperativeAggregateFunction extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43275/
Test FAILed.

JoshRosen commented:
Whoops, looks like I broke something. Will bisect and figure it out tomorrow.

JoshRosen commented:
Okay, I think I might spot the problem: I bet that this was caused by my lifting of the `lazy val cloneBufferAttributes` into the parent class; what's probably happening is that we're now hitting initialization ordering issues. I'll just roll back this change.

davies commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/functions.scala:
```diff
@@ -26,7 +26,7 @@ import org.apache.spark.sql.catalyst.dsl.expressions._
 import org.apache.spark.sql.types._
 
-case class Average(child: Expression) extends AlgebraicAggregate {
+case class Average(child: Expression) extends ExpressionAggregateFunction {
```
`AlgebraicAggregate` sounds better to me (shorter). the other one could be `ImperativeAggregate`

JoshRosen replied:
@rxin and I think that algebraic aggregate is potentially confusing because the [Jim Gray Data Cube paper](http://arxiv.org/pdf/cs/0701155.pdf) already defined the term "algebraic aggregate" to mean something slightly differently than how we're using it here. Technically, an algebraic aggregate (in the Gray terminology) could be implemented via an imperative aggregation function.

davies replied:
I see, thanks. Can we get rid of the ending `Function`, it's anonying to see VeryLongNameInTheSourceCode

JoshRosen replied:
Sure, can do.

JoshRosen committed:
Shorten names by dropping Function suffix.

JoshRosen committed:
Add more override keywords.

JoshRosen commented:
It looks like the old code was actually implicitly depending on some weird `lazy val` initialization ordering properties.

JoshRosen committed:
Fix failing tests with lazy val voodoo.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43303 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43303/consoleFull) for   PR 8973 at commit [`a489d65`](https://github.com/apache/spark/commit/a489d654f54397d52937a81bdfb7d0b651dbd642).

SparkQA commented:
[Test build #43303 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43303/console) for   PR 8973 at commit [`a489d65`](https://github.com/apache/spark/commit/a489d654f54397d52937a81bdfb7d0b651dbd642).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Average(child: Expression) extends ExpressionAggregate `
  * `case class Count(child: Expression) extends ExpressionAggregate `
  * `case class First(child: Expression) extends ExpressionAggregate `
  * `case class Last(child: Expression) extends ExpressionAggregate `
  * `case class Max(child: Expression) extends ExpressionAggregate `
  * `case class Min(child: Expression) extends ExpressionAggregate `
  * `abstract class StddevAgg(child: Expression) extends ExpressionAggregate `
  * `case class Sum(child: Expression) extends ExpressionAggregate `
  * `sealed abstract class AggregateFunction2 extends Expression with ImplicitCastInputTypes `
  * `abstract class ImperativeAggregate extends AggregateFunction2 `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43303/
Test PASSed.

JoshRosen commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala:
```diff
@@ -95,98 +92,192 @@ private[sql] case class AggregateExpression2(
-  def withNewMutableBufferOffset(newMutableBufferOffset: Int): Unit = {
-    mutableBufferOffset = newMutableBufferOffset
-  }
+  def inputAggBufferAttributes: Seq[AttributeReference]
```
@yhuai, why do we need to use different attributes to refer to the aggregation buffer columns when they are embedded inside of a `[grouping columns][partial agg columns]` result? Why can't you just re-use the same attributes? It seems like this should significantly improve the code, but I might be overlooking a reason why we need to do this.

JoshRosen replied:
Ah, forgot: we need them to distinguish between the left and right side when merging two buffers.

yhuai commented:
LGTM. Thank you for working on it! Merging to master.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/interfaces.scala:
```diff
@@ -95,98 +92,192 @@ private[sql] case class AggregateExpression2(
 
-  /** Clones bufferAttributes. */
-  def cloneBufferAttributes: Seq[Attribute]
+  def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): Unit = {
```
Since this method returns `Unit` instead of `ImperativeAggregate`, maybe `setXXX` is a better name than `withXXX`?

JoshRosen replied:
Good point. This is a carry-over from the old code.
